### PR TITLE
Enable flake8-pytest-style

### PR DESCRIPTION
**Pull request by CoolCat467**
This PR enables the flake8-pytest-style ruff rule.

### CODE REVIEW

**Comment by TeamSpen210** [comment]
Setting the [fixture-parentheses](https://docs.astral.sh/ruff/settings/#flake8-pytest-style-fixture-parentheses) option to `false` in `pyproject.toml` will make Ruff use `@pytest.fixture`, and instead warn if empty parentheses are used. It's probably better to be configuring linters to match project conventions, when the existing code is perfectly correct.

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:140]: :tada:

I also like the setting that forces using a sequence of strings in parametrize rather than a comma-separated string...

File: trio/_core/_tests/test_instrumentation.py
Comment (webknjaz) [inline trio/_core/_tests/test_instrumentation.py:258]: People keep forgetting that it's a regex. Could we make that explicit?
```suggestion
        with pytest.raises(ValueError, match="^oops$"):
```

File: trio/_core/_tests/test_io.py
Comment (webknjaz) [inline trio/_core/_tests/test_io.py:322]:
```suggestion
        with pytest.raises(OSError, match="^Bad file descriptor$"):
```

File: trio/_tests/tools/test_mypy_annotate.py
Comment (webknjaz) [inline trio/_tests/tools/test_mypy_annotate.py:16]: :rocket:

File: trio/_tests/test_windows_pipes.py
Comment (webknjaz) [inline trio/_tests/test_windows_pipes.py:50]: Could you turn all the `match=` arg values into full start-to-end matching regexps?

**Comment by CoolCat467** [comment]
Interesting, the pypy-3.9 runner is segfaulting inside pytest from a `from .test_socket import setsockopt_tests` line. Any ideas?

File: src/trio/_tests/test_socket.py
Comment (jakkdl) [inline src/trio/_tests/test_socket.py:838]: It wants you to do this:
```python
with pytest.raises(OSError, match=...) as excinfo:
  ...
assert any(... for arg in excinfo.value.args)
```

Looks like there's a few more instances of PT017 you can fix similarly

File: src/trio/_tests/test_socket.py
Comment (jakkdl) [inline src/trio/_tests/test_socket.py:1136]: .

File: src/trio/_core/_tests/test_run.py
Comment (jakkdl) [inline src/trio/_core/_tests/test_run.py:145]: .

File: src/trio/_core/_tests/test_run.py
Comment (jakkdl) [inline src/trio/_core/_tests/test_run.py:461]: oh derp, I should've thought of this one myself. You can leave this one as a `noqa` though and let me handle it in the multierror PRs

File: src/trio/_core/_tests/test_multierror.py
Comment (jakkdl) [inline src/trio/_core/_tests/test_multierror.py:139]: PT012 is a bit of a pain when `trio` so often tests context managers, so not entirely sure it's worth having enabled. It did catch a couple ugly cases this one time though.

File: src/trio/_tests/test_highlevel_socket.py
Comment (jakkdl) [inline src/trio/_tests/test_highlevel_socket.py:294]:
```python
for code, match in zip((errno.EMFILE, errno.EFAULT, errno.ENOBUFS), ("Out of file descriptors", "attempt to write to read-only memory", "out of buffers"):
  with assert_checkpoints():
    with pytest.raises(OSError, match=match) as excinfo:
      ...
    ...
```
should work?

File: src/trio/_tests/test_socket.py
Comment (jakkdl) [inline src/trio/_tests/test_socket.py:653]: could save this in a string variable and reuse it

File: src/trio/_tests/test_subprocess.py
Comment (jakkdl) [inline src/trio/_tests/test_subprocess.py:343]: this string also makes sense to save in a variable and reuse

**Review by jakkdl — COMMENTED** [review]
A couple nits, but overall great!

Going to be a ton of merge conflicts though :upside_down_face:

**Comment by jakkdl** [comment]
> Interesting, the pypy-3.9 runner is segfaulting inside pytest from a `from .test_socket import setsockopt_tests` line. Any ideas?
> 
> Edit: https://github.com/python-trio/trio/actions/runs/7054572593/job/19203654087

it worked on a rerun... :woman_shrugging:

File: src/trio/_tests/test_highlevel_socket.py
Comment (webknjaz) [reply src/trio/_tests/test_highlevel_socket.py:294]: By the way, it could be useful to have `@pytest.mark.parametrize` or `pytest-subtests` here so that separate checks would show up in the test report distinctly.

**Comment by webknjaz** [comment]
> > Interesting, the pypy-3.9 runner is segfaulting inside pytest from a `from .test_socket import setsockopt_tests` line. Any ideas?
> > Edit: [python-trio/trio/actions/runs/7054572593/job/19203654087](https://github.com/python-trio/trio/actions/runs/7054572593/job/19203654087)
> 
> it worked on a rerun... 🤷‍♀️

I've been observing flaky PyPy 3.9 segfaults on GHA in other projects FTR.

File: src/trio/_tests/test_socket.py
Comment (CoolCat467) [reply src/trio/_tests/test_socket.py:1136]: This one is quite interesting actually. This one is a loop that breaks prematurely only if the kernel responds with an `Out of socket file ids` error.

File: src/trio/_tests/test_subprocess.py
Comment (CoolCat467) [reply src/trio/_tests/test_subprocess.py]: Changed in f89cd0a

File: src/trio/_tests/test_socket.py
Comment (CoolCat467) [reply src/trio/_tests/test_socket.py]: Changed in f89cd0a

File: src/trio/_tests/test_highlevel_socket.py
Comment (CoolCat467) [reply src/trio/_tests/test_highlevel_socket.py]: Changed in f89cd0a. I didn't do the parameterize because it would be a lot of setup code re-running, but if this is important I can still change it.

File: src/trio/_core/_tests/test_run.py
Comment (CoolCat467) [reply src/trio/_core/_tests/test_run.py]: Changed in f89cd0a

File: src/trio/_tests/test_socket.py
Comment (CoolCat467) [reply src/trio/_tests/test_socket.py]: Changed in f89cd0a

File: src/trio/_tests/test_socket.py
Comment (jakkdl) [reply src/trio/_tests/test_socket.py:1132]: hmm, this is a weird test. it loops `_x` until it gets `total//2-1` unless it throws an exception - ~~in which case it'd break and then the test would fail later in the last~~ nowait that's just a print! That print should totally be inside the `except` catcher and `_x` can be renamed to `_`. Probably also a `pragma: no cover`? codecov is acting weird so idk if it's getting covered or not.

Okay, so, it loops until it fails to create a socket - or it hits `total//2`. So supposedly it's not an error if it raises earlier, nor is it a problem if it doesn't raise, which means that `pytest.raises` is not relevant.

It'd maybe be nice to add some comments. Why `pytest.raises` isn't relevant, and maybe also nice to add that EMFILE is "Too many open files" and ENFILE is "File table overflow".

**Review by jakkdl — APPROVED** [review]
Can *probably* ignore the codecov check failing, but it might be nice to make use of it to check out what's not properly getting run - and figure out if it should have `pragma: no cover` or if you messed with some check that now doesn't run anymore.

File: src/trio/_tests/test_socket.py
Comment (CoolCat467) [reply src/trio/_tests/test_socket.py:1132]: Fixed I think with 48b771b4

**Comment by jakkdl** [comment]
I fixed the codecov fails - this PR didn't touch any of those files, but it changed the length of the files which modified the % of misses in the files, which caused it to fail :upside_down_face:

So, unrelated, but:
@agronholm looking at `git blame` it looks like your code for testing `open_tcp_listeners` have had dead code since inception. It always has an `ExceptionGroup` as the `__cause__`:
https://github.com/python-trio/trio/blob/60b9e1423b5f1a80798e214af96926a80c7d94d2/src/trio/_highlevel_open_tcp_listeners.py#L166-L173
but the code testing it (that I removed in my commit) has a branch for if the `__cause__` is not wrapped in an exceptiongroup:
https://github.com/python-trio/trio/blob/master/src/trio/_tests/test_highlevel_open_tcp_listeners.py#L326-L335

I don't really see much any upside to changing the behaviour, but that could be done easily, so I opted to simply remove the redundant test code.

**Comment by jakkdl** [comment]
ah no, it was codecov apparently being smart enough to figure out that the assertion in the test was irrelevant given that the `raises` now has a `match=` ?!

**Comment by jakkdl** [comment]
I would like to merge this unless somebody else wants to review

**Review by TeamSpen210 — APPROVED** [review]
Looks all good. Perhaps we might want to enable the other config too to prevent parentheses in `@pytest.mark.whatever()` also, but not important.

**Comment by CoolCat467** [comment]
> Perhaps we might want to enable the other config too to prevent parentheses in `@pytest.mark.whatever()` also, but not important.

I think this PR is a bit big, if we would like to later I think another PR would be best.

**Comment by CoolCat467** [comment]
Merging, we have two approvals.

**Pull request merged**
